File: web/src/utils/monthlyDrawSchedule.js
export const getNextDrawTimeMillis = (fromTimeMillis = Date.now()) => {
  const date = new Date(fromTimeMillis);
  let year = date.getFullYear();
  let month = date.getMonth(); // 0-indexed (0=Jan, 9=Oct, 11=Dec)
  const day = date.getDate();
  const hours = date.getHours();

  const firstSeasonStartYear = 2026;
  const activeMonths = [9, 10, 11, 0, 1, 2, 3, 4, 5]; // Oct, Nov, Dec, Jan, Feb, Mar, Apr, May, Jun

  let targetYear = year < firstSeasonStartYear ? firstSeasonStartYear : year;
  let targetMonth = month;

  const targetDayThisMonth = targetMonth === 1 ? 28 : 30; // 28th for Feb, 30th for others
  const passedThisMonth = day > targetDayThisMonth || (day === targetDayThisMonth && hours >= 20);

  if (passedThisMonth) {
    targetMonth++;
    if (targetMonth > 11) {
      targetMonth = 0;
      targetYear++;
    }
  }

  while (!activeMonths.includes(targetMonth) || targetYear < firstSeasonStartYear) {
    targetMonth++;
    if (targetMonth > 11) {
      targetMonth = 0;
      targetYear++;
    }
  }

  const targetDay = targetMonth === 1 ? 28 : 30;
  const targetDate = new Date(targetYear, targetMonth, targetDay, 20, 0, 0, 0);
  return targetDate.getTime();
};

File: app/src/main/java/com/ntvelop/mprikiwinners/domain/util/MonthlyDrawSchedule.kt
package com.ntvelop.mprikiwinners.domain.util

import java.util.Calendar
import java.util.TimeZone

object MonthlyDrawSchedule {

    /**
     * Active draw months: October (9), November (10), December (11), January (0), 
     * February (1), March (2), April (3), May (4), June (5). (0-indexed in Calendar)
     */
    private val ACTIVE_SEASON_MONTHS = listOf(
        Calendar.OCTOBER,
        Calendar.NOVEMBER,
        Calendar.DECEMBER,
        Calendar.JANUARY,
        Calendar.FEBRUARY,
        Calendar.MARCH,
        Calendar.APRIL,
        Calendar.MAY,
        Calendar.JUNE
    )

    /**
     * Calculates the next draw timestamp on the 30th of each month (and 28th for February) at 20:00 (8 PM).
     * Season starts October 30, 2026 and runs every month until June 2027, 
     * repeating every October-June season.
     */
    fun getNextDrawTimeMillis(fromTimeMillis: Long = System.currentTimeMillis()): Long {
        val cal = Calendar.getInstance(TimeZone.getDefault()).apply {
            timeInMillis = fromTimeMillis
        }

        val year = cal.get(Calendar.YEAR)
        val month = cal.get(Calendar.MONTH)
        val day = cal.get(Calendar.DAY_OF_MONTH)

        // Check target dates starting from 30 Oct 2026
        val firstSeasonStartYear = 2026

        var targetYear = if (year < firstSeasonStartYear) firstSeasonStartYear else year
        var targetMonth = month

        val drawDayThisMonth = if (targetMonth == Calendar.FEBRUARY) 28 else 30
        val passedThisMonthDraw = day > drawDayThisMonth || (day == drawDayThisMonth && cal.get(Calendar.HOUR_OF_DAY) >= 20)

        if (passedThisMonthDraw) {
            targetMonth++
            if (targetMonth > Calendar.DECEMBER) {
                targetMonth = Calendar.JANUARY
                targetYear++
            }
        }

        // Find the next active month in the Oct-Jun season
        while (!ACTIVE_SEASON_MONTHS.contains(targetMonth) || targetYear < firstSeasonStartYear) {
            targetMonth++
            if (targetMonth > Calendar.DECEMBER) {
                targetMonth = Calendar.JANUARY
                targetYear++
            }
        }

        // Build target calendar date (28th for February, 30th for other months at 20:00)
        val targetCal = Calendar.getInstance(TimeZone.getDefault()).apply {
            clear()
            set(Calendar.YEAR, targetYear)
            set(Calendar.MONTH, targetMonth)
            
            val targetDay = if (targetMonth == Calendar.FEBRUARY) 28 else 30

            set(Calendar.DAY_OF_MONTH, targetDay)
            set(Calendar.HOUR_OF_DAY, 20)
            set(Calendar.MINUTE, 0)
            set(Calendar.SECOND, 0)
            set(Calendar.MILLISECOND, 0)
        }

        return targetCal.timeInMillis
    }
}
